src/Processors/Formats/Impl/NpyOutputFormat.cpp

```diff
@@ -0,0 +1,243 @@
+        case TypeIndex::Float32: numpy_data_type = NumpyDataType('<', 'f', sizeof(Float32)); break;
+        case TypeIndex::Float64: numpy_data_type = NumpyDataType('<', 'f', sizeof(Float64)); break;
+        case TypeIndex::FixedString: numpy_data_type = NumpyDataType('|', 'S', assert_cast<const DataTypeFixedString *>(type.get())->getN()); break;
+        case TypeIndex::String: numpy_data_type = NumpyDataType('|', 'S', 0); break;
+        case TypeIndex::Array: return getNumpyDataType(assert_cast<const DataTypeArray *>(type.get())->getNestedType());
+        default: nested_data_type = type; return false;
+    }
+
+    nested_data_type = type;
+    return true;
+}
+
+void NpyOutputFormat::consume(Chunk chunk)
+{
+    if (!invalid_shape)
+    {
+        num_rows += chunk.getNumRows();
+        const auto & column = chunk.getColumns()[0];
+
+        if (!is_initialized)
+        {
+            initShape(column);
+            is_initialized = true;
+        }
+
+        if (!checkShape(column))
+        {
+            invalid_shape = true;
+            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "ClickHouse doesn't support object types, cannot format ragged nested sequences (which is a list of arrays with different shapes)");
+        }
+    }
+}
+
+void NpyOutputFormat::initShape(const ColumnPtr & column)
+{
+    auto type = data_type;
+    ColumnPtr nested_column = column;
+    while (type->getTypeId() == TypeIndex::Array)
+    {
+        const auto * array_column = assert_cast<const ColumnArray *>(nested_column.get());
+
+        numpy_shape.push_back(array_column->getOffsets()[0]);
+
+        type = assert_cast<const DataTypeArray *>(type.get())->getNestedType();
+        nested_column = array_column->getDataPtr();
+    }
+}
+
+bool NpyOutputFormat::checkShape(const ColumnPtr & column)
+{
+    auto type = data_type;
+    ColumnPtr nested_column = column;
+    int dim = 0;
+    while (type->getTypeId() == TypeIndex::Array)
+    {
+        const auto * array_column = assert_cast<const ColumnArray *>(nested_column.get());
+        const auto & array_offset = array_column->getOffsets();
+
+        for (size_t i = 1; i < array_offset.size(); ++i)
+            if (array_offset[i] - array_offset[i - 1] != numpy_shape[dim])
+                return false;
+
+        type = assert_cast<const DataTypeArray *>(type.get())->getNestedType();
+        nested_column = array_column->getDataPtr();
+        dim += 1;
+    }
+
+    if (type->getTypeId() == TypeIndex::String)
+    {
+        const auto & string_offsets = assert_cast<const ColumnString *>(nested_column.get())->getOffsets();
+        for (size_t i = 0; i < string_offsets.size(); ++i)
+        {
+            size_t string_length = static_cast<size_t>(string_offsets[i] - 1 - string_offsets[i - 1]);
+            numpy_data_type.size = numpy_data_type.size > string_length ? numpy_data_type.size : string_length;
+        }
+    }
+
+    columns.push_back(nested_column);
```

Comment:
It's a bit strage to add it in `checkShape` method. Let's call it in `consume` after all checks